mongodb: 4.0.12 -> 4.2.8

pjjw opened the pull request:
Not strictly an upgrade, but adds a new mongodb-4_2 target with the
current mongodb from that branch.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change
Current mongodb isn't packaged, would like it to be.


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [X] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

pjjw committed:
mongodb: 4.0.12 -> 4.2.8

Not strictly an upgrade, but adds a new mongodb-4_2 target with the
current mongodb from that branch.

Use matching client and server versions in mongodb tests- tests were
using the mongo 3.4 client to connect, and this finally doesn't work
with server 4.2.

Per reviewer suggestion, adding myself as cheetah3 maintainer.

Additionally, reestore comments describing the purpose of the
build-dependencies patch

pjjw commented:
/marvin opt-in
/status needs_reviewer

fgaz commented on pkgs/development/python-modules/cheetah3/default.nix:
```diff
@@ -0,0 +1,19 @@
+  doCheck = false; # Circular dependency
+
+  meta = {
+    homepage = http://www.cheetahtemplate.org/;
```
Quotes around the url https://github.com/NixOS/rfcs/pull/45

pjjw replied:
done

fgaz commented on pkgs/development/python-modules/cheetah3/default.nix:
```diff
@@ -0,0 +1,19 @@
+    description = "A template engine and code generation tool";
+    license = lib.licenses.mit;
+  };
+}
```
Nixpkgs' style guide requires a newline at eof

pjjw replied:
done

fgaz commented on pkgs/servers/nosql/mongodb/mongodb.nix:
```diff
@@ -11,6 +11,17 @@ with stdenv.lib;
 }@args:
 
 let
+  ifdef = if versionAtLeast version "4.2"
```
"ifdef" is not very self-explanatory, I'd change it with something like "pythonVariant"

pjjw replied:
yeah, good suggestion, will do, thanks!

fgaz commented on pkgs/servers/nosql/mongodb/mongodb.nix:
```diff
@@ -11,6 +11,17 @@ with stdenv.lib;
 let
+  ifdef = if versionAtLeast version "4.2"
+    then { python = python38.withPackages (ps: with ps; [ pyyaml cheetah3 psutil setuptools ]);
+            scons = scons;
```
Maybe inherit? I do like the symmetry with the one below though :)

pjjw replied:
this reads a bit clearer to me, though admittedly i am a novice here.

fgaz commented on pkgs/development/python-modules/cheetah3/default.nix:
```diff
@@ -0,0 +1,19 @@
+    homepage = http://www.cheetahtemplate.org/;
+    description = "A template engine and code generation tool";
+    license = lib.licenses.mit;
+  };
```
What about adding yourself as the maintainer of the library? :)

pjjw replied:
done

fgaz reviewed (approved):
The diff looks fine (didn't look at the `.patch`es). Some style suggestions.

pjjw commented:
addressed comments, squashed out the errors i'd made, noticed that github doesn't preserve that history- apologies for that. regardless, PTAL. thanks!

pjjw commented:
/status needs_merger

fgaz commented:
@GrahamcOfBorg build mongodb-4_2

fgaz commented:
Oh, right. It's unfree now.

pjjw commented:
actually just caught an issue with tests- current test is using cross-version clients, and 3.4 to 4.2 doesn't work. quick update inbound.

fgaz reviewed (approved):
I pinged a few more people that were in the maintainer list. If no one replies in a couple days I'll just set this to needs_merge

fgaz commented:
In the meantime please prefix the commit messages with their scope, like you did on the first one

pjjw commented:
sure. mind if i just squash them out now that you've reviewed?

fgaz commented:
That's also fine, yes

fgaz commented:
/status needs_merger
